### PR TITLE
Update toolchains on tioga, lassen, ruby and poodle

### PR DESCRIPTION
**Pull request by adrienbernede**
- [x] Remove Intel 19
- [x] Update to Intel 2023
- [x] ~~Update corona to ROCm 6.0.2~~ -> same error as with tioga, need ROCm 6.1.x , reverted to ROCm 5.7.1
- [x] Update tioga to ROCm 6.1.2
- [x] Remove CUDA 10 job (blueos system default is now 11.2.0)

## Errors

- gcc 11.2.1 + cuda 11.8.0
```
[ 69%] Linking CXX executable ../../bin/managed_array_unit_tests
tmpxft_0001363d_00000000-6_managed_array_tests.cudafe1.stub.c:35:1: error: redefinition of 'const char _ZTSZN4chai12ManagedArrayIdE8allocateEmNS_14ExecutionSpaceERKSt8functionIFvPKNS_13PointerRecordENS_6ActionES2_EEEd_UlS6_S7_S2_E_ []'
tmpxft_0001363d_00000000-6_managed_array_tests.cudafe1.stub.c:35:1: note: 'const char _ZTSZN4chai12ManagedArrayIdE8allocateEmNS_14ExecutionSpaceERKSt8functionIFvPKNS_13PointerRecordENS_6ActionES2_EEEd_UlS6_S7_S2_E_ [125]' previously defined here
tmpxft_0001363d_00000000-6_managed_array_tests.cudafe1.stub.c:35:1: error: redefinition of 'const char _ZTSZN4chai12ManagedArrayIfE8allocateEmNS_14ExecutionSpaceERKSt8functionIFvPKNS_13PointerRecordENS_6ActionES2_EEEd_UlS6_S7_S2_E_ []'
tmpxft_0001363d_00000000-6_managed_array_tests.cudafe1.stub.c:35:1: note: 'const char _ZTSZN4chai12ManagedArrayIfE8allocateEmNS_14ExecutionSpaceERKSt8functionIFvPKNS_13PointerRecordENS_6ActionES2_EEEd_UlS6_S7_S2_E_ [125]' previously defined here
gmake[2]: *** [tests/integration/CMakeFiles/managed_array_tests.dir/build.make:76: tests/integration/CMakeFiles/managed_array_tests.dir/managed_array_tests.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:1724: tests/integration/CMakeFiles/managed_array_tests.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
```

### CODE REVIEW

File: .gitlab/jobs/lassen.yml [hunk 1]
```diff
@@ -47,15 +45,45 @@ clang_12_0_1_libcpp:
...
+    SPEC: "~shared +raja tests=basic %clang@16.0.6 cflags==\"-DGTEST_HAS_CXXABI_H_=0\" cxxflags==\"-stdlib=libc++ -DGTEST_HAS_CXXABI_H_=0\""
+  allow_failure: true
+
+clang_12_0_1_libcpp:
+  extends: .job_on_lassen
+  variables:
+    SPEC: "~shared +raja tests=basic %clang@12.0.1 cflags==\"-DGTEST_HAS_CXXABI_H_=0\" cxxflags==\"-stdlib=libc++ -DGTEST_HAS_CXXABI_H_=0\""
+  allow_failure: true
+
+clang_16_0_6_libcpp:
```
Comment (adayton1) [inline .gitlab/jobs/lassen.yml:60]: This one looks like it's in here twice.

**Review by adayton1 — APPROVED** [review]
LGTM

**Review by liu15 — APPROVED** [review]

**Comment by adayton1** [comment]
> @adayton1 regarding the failure in the gcc 11.2.1 + cuda 11.8 build, do you think I should ignore the failure or wait for a fix?

I'm looking into this again.

**Comment by adayton1** [comment]
I think nvcc preprocesses std::function incorrectly. This isn't the first time I've seen issues with nvcc and the <functional> std library header.

**Comment by adrienbernede** [comment]
@adayton1 I was attempting to understand the issue better, and here is my conclusion:

In `src/chai/ManagedArray.inl` at line 444, 
```cpp
template<typename T>
template<typename Idx>
CHAI_INLINE
CHAI_HOST_DEVICE T& ManagedArray<T>::operator[](const Idx i) const {
  return m_active_pointer[i];
}
```
This operator appeares to be defined for both \_\_host\_\_ and \_\_device\_\_, while there exists an implementation of the class ManagedArray that is only defined for \_\_host\_\_,
In `src/chai/ManagedArray.inl` at line 82,
```cpp
template<typename T>
CHAI_INLINE
CHAI_HOST ManagedArray<T>::ManagedArray(PointerRecord* record, ExecutionSpace space):
[...]
```

I think this explains the warning.


Then, the error itself appears to be related in the following way:
It looks like there is a mismatch between class implementation.
In `` at line ,
```cpp
TEST(ManagedArray, SpaceConstructorCPU)
{
  chai::ManagedArray<float> array(10, chai::CPU);
  ASSERT_EQ(array.size(), 10u);
  array.free();
}
```

I think the intent is for the test to instantiate
```cpp
CHAI_HOST_DEVICE ManagedArray<T>::ManagedArray(
    size_t elems,
    ExecutionSpace space) :
  ManagedArray()
```
defined at line 53 of `src/chai/ManagedArray.inl`, but the error clearly refers to the instance involved in the warning, that is \_\_host\_\_ only (line 84):
```cpp
template<typename T>
CHAI_INLINE
CHAI_HOST ManagedArray<T>::ManagedArray(PointerRecord* record, ExecutionSpace space):
[...]
```

Could it be that `PointerRecord*` and `size_t` are not discernable ? I’m already beyond my limits... ;)

**Comment by adrienbernede** [comment]
@adayton1 You may now merge the PR.

**Pull request merged**